File: code/packages/python/caesar-cipher/tests/test_cipher.py
"""tests/test_cipher.py -- Comprehensive tests for Caesar cipher encrypt/decrypt/rot13.

These tests verify correctness of the core cipher operations including:
- Encrypt / decrypt round-trip for all shifts 0-25
- Case preservation (uppercase stays uppercase, lowercase stays lowercase)
- Non-alphabetic passthrough (digits, punctuation, spaces)
- Empty string handling
- Negative shifts
- Shift wrapping (26, 52, -26 behave as identity)
- ROT13 self-inverse property
- Classic worked example: "HELLO" -> "KHOOR" with shift 3
"""

from __future__ import annotations

import pytest

from caesar_cipher.cipher import decrypt, encrypt, rot13

# ---------------------------------------------------------------------------
# Classic worked example
# ---------------------------------------------------------------------------


class TestClassicExample:
    """The canonical Caesar cipher example: HELLO with shift 3."""

    def test_hello_encrypt(self) -> None:
        assert encrypt("HELLO", 3) == "KHOOR"

    def test_hello_decrypt(self) -> None:
        assert decrypt("KHOOR", 3) == "HELLO"


# ---------------------------------------------------------------------------
# Round-trip: encrypt then decrypt should return the original
# ---------------------------------------------------------------------------


class TestRoundTrip:
    """Encrypting and then decrypting with the same shift must be lossless."""

    @pytest.mark.parametrize("shift", list(range(26)))
    def test_round_trip_all_shifts(self, shift: int) -> None:
        plaintext = "The Quick Brown Fox Jumps Over The Lazy Dog!"
        assert decrypt(encrypt(plaintext, shift), shift) == plaintext

    def test_round_trip_lowercase(self) -> None:
        plaintext = "abcdefghijklmnopqrstuvwxyz"
        assert decrypt(encrypt(plaintext, 17), 17) == plaintext

    def test_round_trip_uppercase(self) -> None:
        plaintext = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        assert decrypt(encrypt(plaintext, 10), 10) == plaintext

    def test_round_trip_mixed(self) -> None:
        plaintext = "Hello, World! 123"
        assert decrypt(encrypt(plaintext, 5), 5) == plaintext


# ---------------------------------------------------------------------------
# Case preservation
# ---------------------------------------------------------------------------


class TestCasePreservation:
    """The cipher must preserve the case of each letter."""

    def test_uppercase_stays_uppercase(self) -> None:
        result = encrypt("ABC", 1)
        assert result == "BCD"
        assert result.isupper()

    def test_lowercase_stays_lowercase(self) -> None:
        result = encrypt("abc", 1)
        assert result == "bcd"
        assert result.islower()

    def test_mixed_case_preserved(self) -> None:
        result = encrypt("AbCdEf", 2)
        assert result == "CdEfGh"
        # Check case of each character individually
        assert result[0].isupper()  # A -> C
        assert result[1].islower()  # b -> d
        assert result[2].isupper()  # C -> E
        assert result[3].islower()  # d -> f
        assert result[4].isupper()  # E -> G
        assert result[5].islower()  # f -> h


# ---------------------------------------------------------------------------
# Non-alphabetic passthrough
# ---------------------------------------------------------------------------


class TestNonAlphaPassthrough:
    """Digits, punctuation, spaces, and other non-alpha chars pass through."""

    def test_digits_unchanged(self) -> None:
        assert encrypt("123456", 5) == "123456"

    def test_punctuation_unchanged(self) -> None:
        assert encrypt("!@#$%^&*()", 10) == "!@#$%^&*()"

    def test_spaces_unchanged(self) -> None:
        assert encrypt("   ", 7) == "   "

    def test_mixed_content(self) -> None:
        result = encrypt("Hello, World! 123.", 3)
        assert result == "Khoor, Zruog! 123."

    def test_newlines_and_tabs(self) -> None:
        assert encrypt("a\nb\tc", 1) == "b\nc\td"

    def test_unicode_passthrough(self) -> None:
        # Non-ASCII characters should pass through unchanged
        assert encrypt("cafe\u0301", 1) == "dbgf\u0301"


# ---------------------------------------------------------------------------
# Empty string
# ---------------------------------------------------------------------------


class TestEmptyString:
    """Empty input should produce empty output."""

    def test_encrypt_empty(self) -> None:
        assert encrypt("", 5) == ""

    def test_decrypt_empty(self) -> None:
        assert decrypt("", 5) == ""

    def test_rot13_empty(self) -> None:
        assert rot13("") == ""


# ---------------------------------------------------------------------------
# Negative shifts
# ---------------------------------------------------------------------------


class TestNegativeShifts:
    """Negative shifts should shift left (same as encrypting with 26-shift)."""

    def test_negative_one(self) -> None:
        assert encrypt("B", -1) == "A"

    def test_negative_three(self) -> None:
        assert encrypt("KHOOR", -3) == "HELLO"

    def test_negative_shift_equals_positive_complement(self) -> None:
        text = "Testing negative shifts"
        assert encrypt(text, -5) == encrypt(text, 21)

    def test_negative_wraps_around(self) -> None:
        assert encrypt("A", -1) == "Z"
        assert encrypt("a", -1) == "z"


# ---------------------------------------------------------------------------
# Shift wrapping
# ---------------------------------------------------------------------------


class TestShiftWrapping:
    """Shifts that are multiples of 26 should be identity; large shifts wrap."""

    def test_shift_zero_is_identity(self) -> None:
        text = "No change expected"
        assert encrypt(text, 0) == text

    def test_shift_26_is_identity(self) -> None:
        text = "Full rotation"
        assert encrypt(text, 26) == text

    def test_shift_52_is_identity(self) -> None:
        text = "Double rotation"
        assert encrypt(text, 52) == text

    def test_shift_negative_26_is_identity(self) -> None:
        text = "Negative full rotation"
        assert encrypt(text, -26) == text

    def test_shift_27_equals_shift_1(self) -> None:
        text = "Wrapping test"
        assert encrypt(text, 27) == encrypt(text, 1)

    def test_large_shift(self) -> None:
        text = "Very large shift"
        assert encrypt(text, 1000) == encrypt(text, 1000 % 26)


# ---------------------------------------------------------------------------
# ROT13
# ---------------------------------------------------------------------------


class TestRot13:
    """ROT13 is shift=13, and applying it twice returns the original."""

    def test_rot13_basic(self) -> None:
        assert rot13("Hello") == "Uryyb"

    def test_rot13_self_inverse(self) -> None:
        text = "The quick brown fox jumps over the lazy dog!"
        assert rot13(rot13(text)) == text

    def test_rot13_is_encrypt_13(self) -> None:
        text = "ROT13 test"
        assert rot13(text) == encrypt(text, 13)

    def test_rot13_all_letters(self) -> None:
        # A-M map to N-Z and vice versa
        assert rot13("ABCDEFGHIJKLM") == "NOPQRSTUVWXYZ"
        assert rot13("NOPQRSTUVWXYZ") == "ABCDEFGHIJKLM"

    def test_rot13_preserves_nonalpha(self) -> None:
        assert rot13("Hello, World! 123") == "Uryyb, Jbeyq! 123"


# ---------------------------------------------------------------------------
# Full alphabet shift
# ---------------------------------------------------------------------------


class TestFullAlphabet:
    """Verify the entire alphabet shifts correctly for a few key shifts."""

    def test_shift_1(self) -> None:
        assert encrypt("abcdefghijklmnopqrstuvwxyz", 1) == "bcdefghijklmnopqrstuvwxyza"

    def test_shift_13(self) -> None:
        assert encrypt("abcdefghijklmnopqrstuvwxyz", 13) == "nopqrstuvwxyzabcdefghijklm"

    def test_shift_25(self) -> None:
        assert encrypt("abcdefghijklmnopqrstuvwxyz", 25) == "zabcdefghijklmnopqrstuvwxy"

File: code/packages/ruby/caesar_cipher/test/test_cipher.rb
# frozen_string_literal: true

require_relative "test_helper"

# ============================================================================
# Tests for CodingAdventures::CaesarCipher encrypt / decrypt / rot13
# ============================================================================
#
# We verify:
#   - Basic encryption with known shift values
#   - Case preservation (uppercase stays uppercase, lowercase stays lowercase)
#   - Non-alphabetic characters pass through unchanged
#   - Round-trip: decrypt(encrypt(text, s), s) == text for all shifts
#   - Edge cases: empty strings, shift = 0, negative shifts, large shifts
#   - ROT13 self-inverse property
# ============================================================================

class TestCipher < Minitest::Test
  CC = CodingAdventures::CaesarCipher

  # --------------------------------------------------------------------------
  # Encryption basics
  # --------------------------------------------------------------------------

  def test_encrypt_hello_shift_3
    # The classic example: HELLO with shift 3 becomes KHOOR.
    #   H(7)  + 3 = K(10)
    #   E(4)  + 3 = H(7)
    #   L(11) + 3 = O(14)
    #   L(11) + 3 = O(14)
    #   O(14) + 3 = R(17)
    assert_equal "KHOOR", CC.encrypt("HELLO", 3)
  end

  def test_encrypt_lowercase
    assert_equal "khoor", CC.encrypt("hello", 3)
  end

  def test_encrypt_mixed_case
    assert_equal "Khoor, Zruog!", CC.encrypt("Hello, World!", 3)
  end

  def test_encrypt_shift_1
    assert_equal "BCD", CC.encrypt("ABC", 1)
  end

  def test_encrypt_shift_25
    # Shift 25 is the same as shift -1: each letter moves one back.
    assert_equal "ZAB", CC.encrypt("ABC", 25)
  end

  def test_encrypt_wrapping_xyz_shift_3
    # X(23)+3=A(0), Y(24)+3=B(1), Z(25)+3=C(2)
    assert_equal "abc", CC.encrypt("xyz", 3)
  end

  def test_encrypt_full_alphabet
    # Shift the entire alphabet by 13 (ROT13).
    plain  = "abcdefghijklmnopqrstuvwxyz"
    cipher = "nopqrstuvwxyzabcdefghijklm"
    assert_equal cipher, CC.encrypt(plain, 13)
  end

  # --------------------------------------------------------------------------
  # Non-alpha passthrough
  # --------------------------------------------------------------------------

  def test_digits_pass_through
    assert_equal "123", CC.encrypt("123", 5)
  end

  def test_punctuation_pass_through
    assert_equal "!@#$%", CC.encrypt("!@#$%", 10)
  end

  def test_spaces_pass_through
    assert_equal "B C", CC.encrypt("A B", 1)
  end

  def test_mixed_content_preserves_non_alpha
    assert_equal "Ifmmp, Xpsme! 123", CC.encrypt("Hello, World! 123", 1)
  end

  # --------------------------------------------------------------------------
  # Case preservation
  # --------------------------------------------------------------------------

  def test_case_preservation
    result = CC.encrypt("AaBbZz", 1)
    assert_equal "BbCcAa", result
  end

  # --------------------------------------------------------------------------
  # Edge cases
  # --------------------------------------------------------------------------

  def test_empty_string
    assert_equal "", CC.encrypt("", 5)
  end

  def test_shift_zero
    assert_equal "HELLO", CC.encrypt("HELLO", 0)
  end

  def test_shift_26_is_identity
    # Shifting by 26 (full alphabet) should return the original text.
    assert_equal "HELLO", CC.encrypt("HELLO", 26)
  end

  def test_shift_52_is_identity
    assert_equal "HELLO", CC.encrypt("HELLO", 52)
  end

  def test_negative_shift
    # shift -1 is equivalent to shift 25
    assert_equal "ZAB", CC.encrypt("ABC", -1)
  end

  def test_negative_shift_large
    # -27 % 26 == 25 in Ruby
    assert_equal "ZAB", CC.encrypt("ABC", -27)
  end

  # --------------------------------------------------------------------------
  # Decryption
  # --------------------------------------------------------------------------

  def test_decrypt_khoor_shift_3
    assert_equal "HELLO", CC.decrypt("KHOOR", 3)
  end

  def test_decrypt_lowercase
    assert_equal "hello", CC.decrypt("khoor", 3)
  end

  def test_decrypt_mixed_case_with_punctuation
    assert_equal "Hello, World!", CC.decrypt("Khoor, Zruog!", 3)
  end

  def test_decrypt_empty_string
    assert_equal "", CC.decrypt("", 7)
  end

  def test_decrypt_shift_zero
    assert_equal "HELLO", CC.decrypt("HELLO", 0)
  end

  # --------------------------------------------------------------------------
  # Round-trip: decrypt(encrypt(text, s), s) == text
  # --------------------------------------------------------------------------

  def test_round_trip_all_shifts
    text = "The Quick Brown Fox Jumps Over The Lazy Dog!"
    (0..25).each do |shift|
      encrypted = CC.encrypt(text, shift)
      decrypted = CC.decrypt(encrypted, shift)
      assert_equal text, decrypted, "Round-trip failed for shift #{shift}"
    end
  end

  def test_round_trip_negative_shift
    text = "Testing negative shifts"
    (-25..-1).each do |shift|
      encrypted = CC.encrypt(text, shift)
      decrypted = CC.decrypt(encrypted, shift)
      assert_equal text, decrypted, "Round-trip failed for shift #{shift}"
    end
  end

  # --------------------------------------------------------------------------
  # ROT13
  # --------------------------------------------------------------------------

  def test_rot13_hello
    assert_equal "URYYB", CC.rot13("HELLO")
  end

  def test_rot13_lowercase
    assert_equal "uryyb", CC.rot13("hello")
  end

  def test_rot13_self_inverse
    # Applying ROT13 twice should return the original text.
    text = "Hello, World! 123"
    assert_equal text, CC.rot13(CC.rot13(text))
  end

  def test_rot13_self_inverse_all_letters
    text = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"
    assert_equal text, CC.rot13(CC.rot13(text))
  end

  def test_rot13_preserves_non_alpha
    assert_equal "Uryyb, Jbeyq! 456", CC.rot13("Hello, World! 456")
  end

  def test_rot13_empty_string
    assert_equal "", CC.rot13("")
  end

  def test_rot13_is_encrypt_with_13
    text = "Some arbitrary text!"
    assert_equal CC.encrypt(text, 13), CC.rot13(text)
  end
end
